src/freenet/support/MultiValueTable.java

```diff
@@ -1,230 +1,260 @@
+ * <p>
+ * This object is thread-safe, it may be read and updated concurrently.
+ * Access rules are similar to collections from {@link java.util.concurrent} package.
  *
  * @author oskar
  */
-
 public class MultiValueTable<K,V> {
-    private Hashtable<K, Vector<V>> table;
-    private int ies;
+    private final Map<K, List<V>> table;
 
     public MultiValueTable() {
-        this(16, 3);
+        table = new ConcurrentHashMap<>();
     }
 
     public MultiValueTable(int initialSize) {
-        this(initialSize, 3);
+        table = new ConcurrentHashMap<>(initialSize);
     }
 
-    public MultiValueTable(int initialSize, int initialEntrySize) {
-        table = new Hashtable<K, Vector<V>>(initialSize);
-        ies = initialEntrySize;
+    /**
+     * Deprecated constructor, please use other variant with only one single size parameter
+     * @param initialSize table initial size
+     * @param unused this parameter was used as initial value collection size, and it is not unused anymore.
+     *               Values collections grow according to Java collection rules.
+     */
+    @Deprecated
+    public MultiValueTable(int initialSize, int unused) {
+        this(initialSize);
     }
 
     public static <K, V> MultiValueTable<K, V> from (K[] keys, V[] values) {
-      if (keys.length != values.length) {
-        throw new IllegalArgumentException(String.format(
-            "keys and values must contain the same number of values, but there are %d keys and %d values",
-            keys.length,
-            values.length));
-      }
-      MultiValueTable<K, V> table = new MultiValueTable<>();
-      for (int i = 0; i < keys.length; i++) {
-        table.put(keys[i], values[i]);
-      }
-      return table;
+        if (keys.length != values.length) {
+            throw new IllegalArgumentException(String.format(
+                "keys and values must contain the same number of values, but there are %d keys and %d values",
+                keys.length,
+                values.length));
+        }
+        MultiValueTable<K, V> table = new MultiValueTable<>(keys.length);
+        for (int i = 0; i < keys.length; i++) {
+            table.put(keys[i], values[i]);
+        }
+        return table;
+    }
+
+    @SafeVarargs
+    public static <K, V> MultiValueTable<K, V> from(K key, V... values) {
+        return from(key, Arrays.asList(values));
+    }
+
+    public static <K, V> MultiValueTable<K, V> from(K key, Collection<V> values) {
+        MultiValueTable<K, V> table = new MultiValueTable<>(1);
+        table.putAll(key, values);
+        return table;
     }
 
     public void put(K key, V value) {
-        synchronized (table) {
-            Vector<V> v = table.get(key);
-            if (v == null) {
-                v = new Vector<V>(ies);
-                table.put(key, v);
+        this.table.compute(key, (k, list) -> {
+            if (list == null) {
+                list = new CopyOnWriteArrayList<>();
```

Comment:
Why the CoW list? If we stick to unmodifiable lists instead (replace the list in the map, instead of altering its values), the implementation becomes a whole lot easier to reason about (once you've read a value lost, its contents remain stable forever) and you don't need to mess with wrapping in `Collections.unmodifiableList()` on every read.

Reply:
`CopyOnWriteArrayList` was picked to allow concurrent iteration and access to be safe.
It also was a continuation of the previous implementation, where underlying data structure was modified in place.

Immutable lists are really better here, but these require some more code to write.
Stable collections may also prevent from various concurrent modification bus.

I pushed new changes containing immutable lists instead of `CopyOnWriteArrayList`.